lib/credentials/applicationTokenCertificateCredentials.ts

```diff
@@ -0,0 +1,137 @@
+      throw new Error("thumbprint must be a non empty string.");
+    }
+    super(clientId, domain, tokenAudience, environment as any, tokenCache);
+
+    this.certificate = certificate;
+    this.thumbprint = thumbprint;
+  }
+
+  /**
+   * Tries to get the token from cache initially. If that is unsuccessfull then it tries to get the token from ADAL.
+   * @returns {Promise<TokenResponse>} A promise that resolves to TokenResponse and rejects with an Error.
+   */
+  public async getToken(): Promise<TokenResponse> {
+    try {
+      const tokenResponse = await this.getTokenFromCache();
+      return tokenResponse;
+    } catch (error) {
+      if (error.message.startsWith(AuthConstants.SDK_INTERNAL_ERROR)) {
+        return Promise.reject(error);
+      }
+      const resource = this.getActiveDirectoryResourceId();
+      return new Promise((resolve, reject) => {
+        this.authContext.acquireTokenWithClientCertificate(
+          resource,
+          this.clientId,
+          this.certificate,
+          this.thumbprint,
+          (error: any, tokenResponse: TokenResponse | ErrorResponse) => {
+            if (error) {
+              return reject(error);
+            }
+            if (tokenResponse.error || tokenResponse.errorDescription) {
+              return reject(tokenResponse);
+            }
+            return resolve(tokenResponse as TokenResponse);
+          }
+        );
+      });
+    }
+  }
+
+  /**
+   * Creates a new instance of ApplicationTokenCertificateCredentials.
+   *
+   * @param clientId  The active directory application client id also known as the SPN (ServicePrincipal Name).
+   * See {@link https://azure.microsoft.com/en-us/documentation/articles/active-directory-devquickstarts-dotnet/ Active Directory Quickstart for .Net}
+   * for an example.
+   * @param {string} certificateStringOrFilePath A PEM encoded certificate and private key OR an absolute filepath to the .pem file containing that information. For example:
+   * - CertificateString: "-----BEGIN PRIVATE KEY-----\n<xxxxx>\n-----END PRIVATE KEY-----\n-----BEGIN CERTIFICATE-----\n<yyyyy>\n-----END CERTIFICATE-----\n"
+   * - CertificateFilePath: **Absolute** file path of the .pem file.
+   * @param domain The domain or tenant id containing this application.
+   * @param options AzureTokenCredentialsOptions - Object representing optional parameters.
+   *
+   * @returns ApplicationTokenCertificateCredentials
+   */
+  public static create(
+    clientId: string,
+    certificateStringOrFilePath: string,
+    domain: string,
+    options: AzureTokenCredentialsOptions
+  ): ApplicationTokenCertificateCredentials {
+    if (
+      !Boolean(certificateStringOrFilePath) ||
+      typeof certificateStringOrFilePath.valueOf() !== "string"
+    ) {
+      throw new Error(
+        "'certificateStringOrFilePath' must be a non empty string."
+      );
+    }
+    let certificate: string = certificateStringOrFilePath;
```

Comment:
```suggestion
    const certificate: string = certificateStringOrFilePath.startsWith("-----BEGIN") ? certificateStringOrFilePath : readFileSync;
```
? We can use const in this case.

Reply:
We won't be able to change the value if the `if` condition is true in the following line.

Reply:
We don't need that line in this case, right?

Reply:
Oh yes! We can do something like this:

```js
if (!certificateStringOrFilePath.startsWith("-----BEGIN")) {
      certificateStringOrFilePath = readFileSync(
        certificateStringOrFilePath,
        "utf8"
      );
    }
    const certificatePattern = /(-+BEGIN CERTIFICATE-+)(\n\r?|\r\n?)([A-Za-z0-9\+\/\n\r]+\=*)(\n\r?|\r\n?)(-+END CERTIFICATE-+)/;
    const matchCert = certificateStringOrFilePath.match(certificatePattern);
```